compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -1,1871 +1,931 @@
+          }
         }
       }
-
-      val cond = renderCondition(treeCondStrategy)
-      val res  = codegen._asInstanceOf(testedBinder, nextBinderTp)
-
-      // is this purely a type test, e.g. no outer check, no equality tests (used in switch emission)
-      //def isPureTypeTest = renderCondition(pureTypeTestChecker)
-
-      def impliesBinderNonNull(binder: Symbol): Boolean =
-      // @odersky: scalac is able to infer in this method that nonNullImpliedByTestChecker.Result,
-      // dotty instead infers type projection TreeMakers.this.TypeTestTreeMaker.TypeTestCondStrategy#Result
-      // which in turn doesn't typecheck in this method. Can you please explain why?
-      // dotty deviation
-        renderCondition(nonNullImpliedByTestChecker(binder)).asInstanceOf[Boolean]
-
-      override def toString = "TT" + ((expectedTp, testedBinder.name, nextBinderTp))
-    }
-
-    // need to substitute to deal with existential types -- TODO: deal with existentials better, don't substitute (see RichClass during quick.comp)
-    case class EqualityTestTreeMaker(prevBinder: Symbol, subpatBinder: Symbol, patTree: Tree, override val pos: Position) extends CondTreeMaker {
-      val nextBinderTp = patTree.tpe & prevBinder.info
-      val nextBinder = if (prevBinder eq subpatBinder) freshSym(pos, nextBinderTp) else subpatBinder
-
-      // NOTE: generate `patTree == patBinder`, since the extractor must be in control of the equals method (also, patBinder may be null)
-      // equals need not be well-behaved, so don't intersect with pattern's (stabilized) type (unlike MaybeBoundTyped's accumType, where it's required)
-      val cond = codegen._equals(patTree, prevBinder)
-      val res  = ref(prevBinder).ensureConforms(nextBinderTp)
-      override def toString = "ET" + ((prevBinder.name, patTree))
+      new ElimRedundant(Map())(plan)
     }
 
-    case class AlternativesTreeMaker(prevBinder: Symbol, var altss: List[List[TreeMaker]], pos: Position) extends TreeMaker with NoNewBinders {
-      // don't substitute prevBinder to nextBinder, a set of alternatives does not need to introduce a new binder, simply reuse the previous one
-
-      override private[TreeMakers] def incorporateOuterRebinding(outerSubst: Rebindings): Unit = {
-        super.incorporateOuterRebinding(outerSubst)
-        altss = altss map (alts => propagateRebindings(alts, rebindings))
-      }
-
-      def chainBefore(next: Tree)(codegenAlt: Casegen): Tree = {
-        /*atPos(pos)*/{
-          // one alternative may still generate multiple trees (e.g., an extractor call + equality test)
-          // (for now,) alternatives may not bind variables (except wildcards), so we don't care about the final substitution built internally by makeTreeMakers
-          val combinedAlts = altss map (altTreeMakers =>
-            ((casegen: Casegen) => combineExtractors(altTreeMakers :+ TrivialTreeMaker(casegen.one(Literal(Constant(true)))))(casegen))
-            )
-
-          val find
```

Comment:
it should also check if initializer refers to vars:

```
case class Box[T](var a: T)

x match {
  case x @ Box(t) =>  // this can be filtered by checking that t is not synthetic
   x.a = null;
   t
  case Box(Box(t)) =>  // this can't, as the breakage happens through an alias to the outer
   x.asInstanceOf[Box].a = null;
   t
}